lib/data_schema.ex

```diff
@@ -918,10 +979,59 @@ defmodule DataSchema do
     Map.put(map, field, item)
   end
 
-  # This just lets us use either a module name for the data type OR a one arity fn.
-  defp call_cast_fn({module, fun, args}, value), do: apply(module, fun, [value | args])
-  defp call_cast_fn(module, value) when is_atom(module), do: module.cast(value)
-  defp call_cast_fn(fun, value) when is_function(fun, 1), do: fun.(value)
+  defp call_cast_fn({module, fun, args}, value, original_field) do
+    try do
+      apply(module, fun, [value | args])
+    rescue
+      error ->
+        message = error_message(original_field, error)
+        reraise DataSchema.CastFunctionError, [message: message], __STACKTRACE__
+    end
+  end
+
+  defp call_cast_fn(module, value, original_field) when is_atom(module) do
+    try do
+      module.cast(value)
+    rescue
+      error ->
+        message = error_message(original_field, error)
+        reraise DataSchema.CastFunctionError, [message: message], __STACKTRACE__
+    end
+  end
+
+  defp call_cast_fn(fun, value, original_field) when is_function(fun, 1) do
+    try do
+      fun.(value)
+    rescue
+      error ->
+        message = error_message(original_field, error)
+
+        reraise DataSchema.CastFunctionError, [message: message], __STACKTRACE__
+    end
+  end
```

Comment:
Happy with not having a final fallback match as long we get a guarantee that we will always match on the earlier matches of call_cast_fn/3